### ISSUE TITLE
Regarding LTI 1.3 Canvas

### ISSUE THREAD

**Issue by Rohi145** [IMSGlobal/lti-1-3-php-library#26]
Hi,
We have a doubt regarding LTI 1.3 in canvas as follows:

1. In LTI 1.1, we can identify valid instructor tool launch using consumer key and secret key. So, how to identify valid instructor launch in LTI 1.3 before open id JWT generation?
2. Can we validate deep link request using oauth2 access token verification?

**Comment by viveklm** [comment]
@Rohi145: Did you resolved your issue, I am facing this issue. when launching deep link 

{"errors":{"jwt":[{"attribute":"jwt","type":"JWT verification failure","message":"JWT verification failure"}]}} 

Getting following error not sure what value and where to set private_key_file??

@MartinLenord : I would greatly appreciate it if you kindly give me some feedback.

**Comment by MartinLenord** [comment]
Hey @Rohi145 
Sorry for the slow reply. 
1. If i understand the question correctly, the signature on the `id_token` validates that the request was signed by Canvas. To know who the request was signed for, you must look at the `deployment_id`. I you don't recognise the `deployment_id`, you should allow the user to prove who they are (e.g. logging in as an administrator) then you can link the `deployment_id` to your own account.
2. Deep linking is a message launch not a service call, meaning the request will be an OpenID Connect request like a standard resource launch

@viveklm if you are looking for the config, it lives here https://github.com/IMSGlobal/lti-1-3-php-library/blob/master/src/game_example/db/configs/example.json the private keys that are referenced liv in https://github.com/IMSGlobal/lti-1-3-php-library/tree/master/src/game_example/db

**Comment by Rohi145** [comment]
@MartinLenord ,

> Hey @Rohi145
> Sorry for the slow reply.
> 
> 1. If i understand the question correctly, the signature on the `id_token` validates that the request was signed by Canvas. To know who the request was signed for, you must look at the `deployment_id`. I you don't recognise the `deployment_id`, you should allow the user to prove who they are (e.g. logging in as an administrator) then you can link the `deployment_id` to your own account.
-----Before id_token generation when login is initiated, we get some parameters from respective LMS (ex.login_hint, issuer, etc.).We need to know during this time, which parameter is used to help for valid instructor validation before OIDC launch.

**Comment by MartinLenord** [comment]
The initial OIDC launch is unsigned. You will be given the issuer that can be used along with the url used to login to identify the client id and url to return to. You cannot validate the user at this point as no user information is passed, also this launch isn't signed so you cannot use the information in it to do any actions except redirect back to the platform.

Verifying the user must be done on the id_token launch as that request is signed

**Comment by Rohi145** [comment]
@MartinLenord 

Thanks for info,

Means as of now, In LTI 1.3 we need to authenticate the user from client id and deployment id?
Because consumer key is not part of LTI 1.3.

**Comment by MartinLenord** [comment]
Yes, you will have issuer ->client_id -> deployment_id to identify a user

**Comment by Rohi145** [comment]
Hello, Martin
I hope you are doing well.
After implementing LTI advantage in BB, moodle, D2l, Now I am trying to integrate in canvas:
I have a few questions please guide me about that.
1. Is LTI advantage in canvas support multiple resource creation?
2.what are the limitations in canvas for LTI advantage?

Please guide me asap.
Thanks.

**Comment by MartinLenord** [comment]
Hi @Rohi145
Canvas does not currently support returning multiple content items in deep linking
They also do not correctly support custom parameters in deep linking

**Comment by eagle-13** [comment]
Hi, Martin. Nice to know you. 
I am encountered the error while integrating this example with canvas. Canvas error is like this:
response_code: 500
request_id: f04ccfc1-e8ef-4b10-8fd3-5380adb6da5d
session_id: ceff9b8a92767421033527825eed644c
meta_headers: ''''
format: !ruby/object:Mime::Type
 synonyms:
 - text/x-json
 - application/jsonrequest
 symbol: :json
 string: application/json
 hash: 237643698634756473
HTTP_ACCEPT: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
HTTP_ACCEPT_ENCODING: gzip, deflate
HTTP_HOST: 192.168.34.43:3000
HTTP_REFERER: http://192.168.34.43:3000/
HTTP_USER_AGENT: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML,
 like Gecko) Chrome/86.0.4240.198 Safari/537.36
PATH_INFO: "/api/lti/authorize"
QUERY_STRING: "?client_id=10000000000001&login_hint=535fa085f22b4655f48cd5a36a9215f64c062838&lti_message_hint=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ2ZXJpZmllciI6IjNhMGMxMGMzNmNlMzMwYzc1ZWQ1Y2M4MjU0YTdiOTA3OWJlZThjMzY5ODZmZjMzNjBjZWYyOTUwNmE1MmI5YmI5MDc1MWRiOWY3MjZhMzJiZjVlOWZkMjYxMmRiZjk2MGM2MDQyOWFlNmUxN2EyYTUwMGVjZGU5NTljNDIzOGUxIiwiY2FudmFzX2RvbWFpbiI6IjE5Mi4xNjguMzQuNDM6MzAwMCIsImNvbnRleHRfdHlwZSI6IkNvdXJzZSIsImNvbnRleHRfaWQiOjEwMDAwMDAwMDAwMDAxLCJleHAiOjE2MDc2MDc5NDd9.fRSp8wDTmQ9-SmFbE5nN-KVTrh9h7se-BcoEMXbe6_s&nonce=nonce-5fd225618f5d95.16445852&prompt=none&redirect_uri=http%3A%2F%2F192.168.34.44%3A8005%2Fgame.php&response_mode=form_post&response_type=id_token&scope=openid&state=state-5fd225618f5ba3_76715108"
REQUEST_METHOD: GET
REQUEST_PATH: "/api/lti/authorize"
REQUEST_URI: http://192.168.34.43:3000/api/lti/authorize?client_id=10000000000001&login_hint=535fa085f22b4655f48cd5a36a9215f64c062838&lti_message_hint=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ2ZXJpZmllciI6IjNhMGMxMGMzNmNlMzMwYzc1ZWQ1Y2M4MjU0YTdiOTA3OWJlZThjMzY5ODZmZjMzNjBjZWYyOTUwNmE1MmI5YmI5MDc1MWRiOWY3MjZhMzJiZjVlOWZkMjYxMmRiZjk2MGM2MDQyOWFlNmUxN2EyYTUwMGVjZGU5NTljNDIzOGUxIiwiY2FudmFzX2RvbWFpbiI6IjE5Mi4xNjguMzQuNDM6MzAwMCIsImNvbnRleHRfdHlwZSI6IkNvdXJzZSIsImNvbnRleHRfaWQiOjEwMDAwMDAwMDAwMDAxLCJleHAiOjE2MDc2MDc5NDd9.fRSp8wDTmQ9-SmFbE5nN-KVTrh9h7se-BcoEMXbe6_s&nonce=nonce-5fd225618f5d95.16445852&prompt=none&redirect_uri=http%3A%2F%2F192.168.34.44%3A8005%2Fgame.php&response_mode=form_post&response_type=id_token&scope=openid&state=state-5fd225618f5ba3_76715108
SERVER_NAME: 192.168.34.43
SERVER_PORT: ''3000''
SERVER_PROTOCOL: HTTP/1.1
REMOTE_ADDR: 192.168.34.44
path_parameters: ''{:format=>"json", :controller=>"lti/ims/authentication", :action=>"authorize"}''
query_parameters: ''{"client_id"=>"10000000000001", "login_hint"=>"535fa085f22b4655f48cd5a36a9215f64c062838",
 "lti_message_hint"=>"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ2ZXJpZmllciI6IjNhMGMxMGMzNmNlMzMwYzc1ZWQ1Y2M4MjU0YTdiOTA3OWJlZThjMzY5ODZmZjMzNjBjZWYyOTUwNmE1MmI5YmI5MDc1MWRiOWY3MjZhMzJiZjVlOWZkMjYxMmRiZjk2MGM2MDQyOWFlNmUxN2EyYTUwMGVjZGU5NTljNDIzOGUxIiwiY2FudmFzX2RvbWFpbiI6IjE5Mi4xNjguMzQuNDM6MzAwMCIsImNvbnRleHRfdHlwZSI6IkNvdXJzZSIsImNvbnRleHRfaWQiOjEwMDAwMDAwMDAwMDAxLCJleHAiOjE2MDc2MDc5NDd9.fRSp8wDTmQ9-SmFbE5nN-KVTrh9h7se-BcoEMXbe6_s",
 "nonce"=>"nonce-5fd225618f5d95.16445852", "prompt"=>"none", "redirect_uri"=>"http://192.168.34.44:8005/game.php",
 "response_mode"=>"form_post", "response_type"=>"id_token", "scope"=>"openid", "state"=>"state-5fd225618f5ba3_76715108"}''
request_parameters: "{}"
exception_message: Nil is not a valid JSON source.
hostname: ubuntu
pid: 104750
', 'TypeError') RETURNING "id"  [development:1 master]

Finally, At page, it displays error:
while(1);{"errors":[{"message":"An error occurred.","error_code":"internal_server_error"}],"error_report_id":753}

What' matter?
How to fix it?
Please help me asap.
Thanks.
